Usethis 2.0.0 (#492)
tidyverse/tidyverse.org: content/blog/usethis-2-0-0/index.Rmd

  author: Kiana Bosman
categories: [package] 
tags: [devtools,r-lib,usethis]
---

We're ecstatic to announce the release of usethis v2.0.0 ([usethis.r-lib.org](https://usethis.r-lib.org/)).
usethis is a package that facilitates interactive workflows for R project creation and development.
It's mostly focussed on easing day-to-day package development, but many of its functions are also useful for non-package projects.

You can install usethis from CRAN with:

```{r, eval = FALSE}
install.packages("usethis")
```

This is a major release, involving lots of change under the hood, and yet there are no big, brand new features.
Instead, this release is about making more things "just work" the way you expect.
Below, we list many functions and arguments that have been deprecated, because we've been able to deliver the same functionality (actually, a bit more) with a smaller and simpler user interface.

This blog post hits a few highlights, mostly relating to Git and GitHub functionality:

* We've switched to a new package, gert, for Git operations. This resolves some
  long-standing difficulties around credential-finding and provides more
  reliable support for both HTTPS and SSH remotes.

> * Anything that works for github.com should now work for any GitHub Enterprise

Reviewer: Make it a link?

We've completely refactored the "GitHub host" logic in usethis and GHE should be fully supported now.
In an existing repo, usethis consults the configured Git remotes, filters for remotes that smell like a GitHub deployment, and deduces the target `host`.
As a result, we've deprecated the `host` argument in several functions:

* `use_github_links()`
* `use_github_labels()`, `use_tidy_labels()`
* `use_github_release()`

In `use_github()` and `create_from_github()`, we still have a `host` argument, but there are also other ways to specify the `host`:

* usethis no longer inserts its own opinion about the default `host`. This
  means we no longer get in the way of the existing default behaviour of the gh
  package, which is to consult the `GITHUB_API_URL` environment variable, if it
  is set.
* A couple functions now accept a full URL as the `repo_spec` and, if you do
  that, we discover the `host` from the URL.

#### Give me your full URLs!

The last point above is a nice quality-of-life improvement even when working on github.com.
It means you can copy a URL straight from your browser and, as long as it points somewhere within the target repo, all will be well.

> If you decide to clone Matt Lincoln's clipr package, while perusing its issues, you can provide the current browser URL to `create_from_github()`, i.e. you don't have to trim both ends to isolate `mdlincoln/clipr`.

Reviewer: ```suggestion
For example, if you decide to clone Matt Lincoln's clipr package while perusing its issues, you can just copy the url directly from your browser:
```

Together, this means usethis no longer needs any explicit PAT management and finishes explaining why so many credential-, token-, and host-related functions and arguments have been deprecated.
This is also a major reason why GitHub Enterprise "just works" now.

The new connection to the system-specific Git credential store also means we no longer need to set `GITHUB_PAT` in our `.Renviron` startup files.
It is a better security practice anyway to avoid storing such secrets in plain text, whenever possible.

```
BLAH=foo
GITHUB_PAT=xyz  # <-- WE SUGGEST YOU REMOVE THIS LINE FROM .Renviron
WUT=yo
OTHER_WEB_SERVICE=super-secret-very-powerful-token
```

The `host`-specific PAT is now retrieved from the Git credential store upon first need.
Note that the PAT **is** still cached in an environment variable for reuse during the remainder of the current R session.

Our full recommendations for getting and configuring a PAT are in the new vignette [Managing Git(Hub) Credentials](https://usethis.r-lib.org/articles/articles/git-credentials.html).

### Git protocol

When usethis configures a new Git remote, it must choose a protocol, either HTTPS or SSH.

> The new default in this situation is HTTPS, because we think HTTPS is a better choice, for most users and projects (rationale is above).

Reviewer: ```suggestion
The new default in this situation is HTTPS, because as we explained above, we think HTTPS is the best choice for most users and projects.
```

```{r}
git_protocol()
```

> In relevant functions, such as `use_github()`, you can still specify the `protocol` directly.

Reviewer: I think you could just delete this line

And, as always, you can specify the protocol for a session with `use_git_protocol()` or for all sessions via the `usethis.protocol` option.
Those who prefer SSH may want to set this option in `.Rprofile` going forward.

## Pull request helpers

The team that maintains the tidyverse and r-lib packages makes heavy use of GitHub pull requests for managing internal and external contributions.
The `pr_*()` family of functions supports pull request workflows, for maintainers and contributors.

> This family has gained a couple of functions and some existing functions are more helpful.
> 
> New pull request functions:

Reviewer: ```suggestion
This family has gained a couple of new functions and some improvements to existing functions:

```

* `pr_resume()` resumes work on an existing local PR branch. It can be called
  argument-less, to select a branch interactively.
  
    ```{r, eval = FALSE}
    > pr_resume()
    ℹ No branch specified ... looking up local branches and associated PRs
    Which branch do you want to checkout? (0 to exit) 
    
    1: avalcarcel9-add_use_author --> #833 ('@avalcarcel9'): Add use author
    2:                 latex-hell
    3:                     holder
    4:            patch-for-withr
    
    Selection: 
    ```
  
* `pr_forget()` abandons a PR you initiated locally or fetched from GitHub. It
  only does local clean up and will never delete a remote branch or close a PR.
  

> `pr_fetch()` and `pr_view()` also present an interactive choice, when the target PR is not specified or implied.

Reviewer: ```suggestion
* `pr_fetch()` and `pr_view()` also present an interactive choice, when the target PR is not specified or implied.
```

Reviewer: And move to be second bullet?

## Other goodies

Here's a selection of other new features:

  * A default Git branch named `main` now works. `git_branch_default()` is a new
    function that tries to discover the default branch from the local or remote
    Git repo. Internally, it is used everywhere that we previously assumed a
    default branch named `master`.

>   * `use_github_pages()` and `use_tidy_pkgdown()` are great for turning on

Reviewer: Probably shouldn't promote `use_tidy_pkgdown()` because it's our workflow?

Author: In this case, `use_tidy_pkgdown()` is both our workflow AND what I would recommend for other maintainers as well. `use_tidy_pkgdown()` is `use_pkgdown() + use_github_pages() + use_github_action("pkgdown") + use_pkgdown_url()`.

Not sure what to do about that, if anything. I see your point, i.e. where other "tidy" functions often do something that is overkill or inappropriate for others. But it's not the case here.

Author: In any case, it's not really a blog post problem but potentially a function naming problem.

Reviewer: Yeah, feels like we might've given the wrong name. Maybe should be `use_github_pkgdown()` or similar?

Author: https://github.com/r-lib/usethis/issues/1308

    GitHub Pages and for using GitHub Actions to build and deploy a pkgdown
    site.
  * usethis knows that RStudio >= 1.3 stores user preferences in a file, which
    means that `use_blank_slate()` can be used to opt in to the "never save to
    or restore `.RData`" lifestyle, globally, for a user.
  * `browse_package()` and `browse_project()` are new additions to the
    `browse_*()` family that let the user choose from a list of URLs derived
    from local Git remotes and DESCRIPTION (local or possibly on CRAN).
    
    ```{r, eval = FALSE}
    > browse_package()
    Which URL do you want to visit? (0 to exit) 
    
    1: https://github.com/r-lib/usethis        ('origin' remote)
    2: https://github.com/avalcarcel9/usethis  ('avalcarcel9' remote)
    3: https://usethis.r-lib.org               (URL field in DESCRIPTION)
    4: https://github.com/r-lib/usethis        (URL field in DESCRIPTION)
    5: https://github.com/r-lib/usethis/issues (BugReports field in DESCRIPTION)
    
    Selection: 
    ```
  * The `use_*_license()` functions have gotten a general overhaul and also now

>     work for projects, not just for packages.

Author: @Hadley sorry I got very caught up in describing all the Git/GitHub changes 😂. Would you prefer that this have a subsection?

Reviewer: I don't think it needs its own section, but it would be useful to make a bit more prominent, and mention https://r-pkgs.org/license.html.
